### ISSUE TITLE
Do we want to let plugins make pre-built analysis components in 5.0?

### ISSUE THREAD

**Issue by nik9000** [elastic/elasticsearch#19814]
We removed support for allowing plugins to register pre-built analysis components in #14355. Do we want that back or should we explicitly declare that caching is the responsibility of the analysis provider implemented by the plugin? If we don't want to allow plugins to make pre-built analysis components, what is our rationale for doing it in core? It'd be nice to have that explicitly documented somewhere. I'm willing to translate any comments in this issue into whatever PR we need.

**Comment by jpountz** [comment]
Hmm I am confused: does it mean that for instance Lucene's JapaneseAnalyzer is not exposed anymore?

**Comment by imotov** [comment]
@jpountz it is exposed but if you run elasticsearch with kuromoji plugin and create 100 indices with no settings. You get this:

![screen shot 2016-08-04 at 5 50 26 pm](https://cloud.githubusercontent.com/assets/655851/17419547/f932d38e-5a6b-11e6-9e04-7f93693bdac4.png)

Basically, you get 100 Kuromoji Factories of different kinds, 100 Fingerprint analyzers and 100 Serbian Normalization Filter Factories. At the same time you get only one SnowballAnalyzer, for example.

**Comment by s1monw** [comment]
the difference here is that we have some analyzers that are shared across indices but those can't be configured. The Kuromoji factories where always like this since they are tokenizer, tokenfilter and char filter factories. They where always per index never prebuilt. This only applied to analyzers. We have a set of pre-built ones in core which I think is enough we don't need to expose this to plugins.

**Comment by imotov** [comment]
Ok, I did a bit more digging. So, basically this is what I've done. I created two one-node clusters with all default settings. One cluster with v2.3.4 node with kuromoji plugin installed and another cluster was just a node with today's master that I started from the kuromoji plugin directory using gradle run. I could retry it with a packaged installation but I don't think it would make any difference. After starting each node I ran a script that created 100 indices with default settings:

```
for i in {1..100}; do curl -XPUT localhost:9200/test-$i; done
```

Then I created a heap dump for each node and analyzed them in YourKit. This is the screenshot for 2.3.4:

![screen shot 2016-08-04 at 8 15 09 pm](https://cloud.githubusercontent.com/assets/655851/17422492/8ab753de-5a80-11e6-8774-04ddddcaec20.png)
![screen shot 2016-08-04 at 8 33 32 pm](https://cloud.githubusercontent.com/assets/655851/17422748/ca21c854-5a82-11e6-8f9c-2d627159a8a8.png)

And this is the picture for the current master:

![screen shot 2016-08-04 at 8 14 55 pm](https://cloud.githubusercontent.com/assets/655851/17422496/a25c8144-5a80-11e6-80cf-9e6ccbf6a1de.png)
![screen shot 2016-08-04 at 8 37 29 pm](https://cloud.githubusercontent.com/assets/655851/17422813/5a1a2a96-5a83-11e6-981e-ea3780a10c20.png)

I think removal of [KuromojiIndicesAnalysis](https://github.com/elastic/elasticsearch/pull/14355/files#diff-20f0ed6d04eea4b1d70143e98326e68cL44) is the reason why this is happening. 

As you can see in 2.3.4 only JapaneseStopTokenFilterFactory is created 100 times and I think this is because we forgot to register its prebuilt version in KuromojiIndicesAnalysis, which looks like a bug in 2.3.4.

So, I think @nik9000's and mine original question was "Is there a way to do something similar in master?"

**Comment by s1monw** [comment]
There are many issues with these pre-bulid analyzers/filters/tokenizers. Some have different behavior depending on the version the index was created. Caching is very very difficult here and we reduced it to the core components here. I really think we simplified this for our users and don't force them to make any difficult decisions but rather trade it for the fact that we create a new instance per index for the analyzers coming from a plugin. 

I am convinced that safety first is important here, if we mess up the caching it means the index is corrupt. I think it's ok to have extra instances based on this concern. 
Yet, I do think there might be options to cache by version but I wonder if that is worth the trouble?

**Comment by s1monw** [comment]
@imotov @nik9000 I reference this from a commit here https://github.com/s1monw/elasticsearch/commit/35ee27ec9b8c931b229e3b6d5041489f82c6bbf4 I think we can do stuff like this but we have to keep it internal. Don't complicate the interface for the optimization it's crucial to the success of our plugin APIs.

**Comment by s1monw** [comment]
I just had an offline conversation with @nik9000 @imotov and we concluded that it might be the best option to simplify AnalysisService to just hold analyzers and trash the Tokenizer, TokenFilter and CharFilters once it's fully instantiated. The only consumer of those are tests, custom analyzer building (which happens during construction of AnalysisService) as well as the `TransportAnalyzeAction`. In `TransportAnalyzeAction` it's really just needed to build custom analyzers on the fly and I think we shouldn't optimize for this. Rather go the slow path and create the analysis service again configuring a Custom analzyer than complicating the memory footprint and caching of core service. I think we should fix this rather than add caching or expand plugin APIs

**Comment by nik9000** [comment]
> we shouldn't optimize for this

I've opened #19827 for this.

> simplify AnalysisService

Opened #19828 with the intention of doing that after #19827.

I've also opened #19830 to investigate removing some of the pre-built analysis components, some of which really don't seem needed.


After opening those issues, I think this issue can now be solved by adding documentation to `AnalysisPlugin` explaining that you are on your own for caching though, in general, you should avoid caching. I'll push a patch adding such documentation and close this issue.
